Turn off desktop notifications by default for new users.

nicholasbs opened the pull request:
Previously, it was a really unpleasant to join an active Zulip realm, because the moment you joined, you were bombarded with dozens of desktop notifications for new messages to every stream you were subscribed to by default. This isn't a theoretical concern: We've on-boarded hundreds of Zulip users at the Recurse Center, and since we have an active realm and lots of default streams, we have to have extra instructions to tell every new user to turn these two settings off as soon as they first log in.

smarx commented:
*Automated message from Dropbox CLA bot*

@nicholasbs, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented:
Personally I agree that we need to do something to change these settings because they're definitely wrong for an established realm and it's annoying for one of your first things to do to be changing the notifications settings, but I think we should be thoughtful here, because IIRC these defaults were set that way to address an issue with newly created realms where users weren't noticing that other people were sending messages.  

I'd be interested in other folks' thoughts on how to solve the problem for new teams effectively here; e.g. maybe we should flag notification settings as material we more strongly encourage users to read or see (e.g. making it its own page in the gear menu separate from the rest of "settings")

(@nicholasbs you should linewrap your commit message content to match the zulip commit message style)

lfranchi commented:
Yeah in general I think this is the right move. We switched to this default but IMO experience shows that tons of people have found this quite annoying.

nicholasbs commented:
I can see how these notifications make sure people don't miss messages early on, but default notifications for everything seems like the wrong way to fix this problem, even for new realms.  Having this on by default encourages a pattern that (I assume) very few Zulip users actually follow once they understand the product (though you'd know better than me; I can only speak for RC's community, and perhaps we're an outlier).

A better approach might be more aggressive email notifications early on, a better tutorial/on-boarding, or even one-time desktop notifications for new users (e.g., to let them know there's a new message and point them towards the notification settings). But I think these are all largely orthogonal to this change.

That said, assuming you disagree with the above, either of the following would solve the current problem for RC:

* A realm-level setting to set the default value for these two settings.

* The ability to set these settings when you invite new users (e.g., two checkboxes in the dialog that comes up after you select, "Invite coworkers" from the gear menu).

timabbott commented:
Yeah honestly I'd be happy to just merge this and think about how to advertise the notification settings better in a new issue about adjusting the onboarding emails.  I'll let this sit a day for further commentary though since it is a significant shift.

I think we should put a note about this into a new changelog.md file that can be used to track major changes that folks might want to read about when upgrading.

@nicholasbs I assume you've tested this change?

timabbott commented:
I guess it's probably worth including in this at least a small change to `templates/zerver/emails/followup/day1.text` item 3 explaining briefly something about how by default @-mentions and PMs result in desktop notifications, sounds, and "missed message" emails but that you can adjust these settings either for individual streams on that stream's settings page or for the new stream default on the overall settings page's notifications section.

nicholasbs commented:
I tested this manually by creating a new user via `manage.py` and verifying that those two checkboxes were unchecked by default for that user in their settings. Is there any additional testing I should do? I couldn't think of a meaningful automated test to write.

I'm happy to take a pass at updating the followup email templates (I assume the HTML one would need to be updated as well) and add a changelog, though I'm busy tonight and traveling the next few days, so that likely won't happen until early next week.

timabbott commented:
You should probably create a new user via the actual signup flow and do the
same checkbox test, since it is a different codepath.

If you haven't done this before in the dev VM, the confirmation emails just
get printed out to the manage.py console.

I think early next week is a fine timeline for addressing the rest.


        -Tim Abbott

On Wed, Oct 7, 2015 at 7:22 PM, Nicholas Bergson-Shilcock <
notifications@github.com> wrote:

> I tested this manually by creating a new user via manage.py and verifying
> that those two checkboxes were unchecked by default for that user in their
> settings. Is there any additional testing I should do? I couldn't think of
> a meaningful automated test to write.
>
> I'm happy to take a pass at updating the followup email templates (I
> assume the HTML one would need to be updated as well) and add a changelog,
> though I'm busy tonight and traveling the next few days, so that likely
> won't happen until early next week.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zulip/zulip/pull/141#issuecomment-146363307>.
>

nicholasbs commented:
@timabbott Ok, I think this is all set:

* I've created a new user (by clicking "Register" and then using the activation link printed to the console) and verified that the two checkboxes under settings were unchecked.
* I've added a `changelog.md`.
* I've added some copy to `day1.html` and `day1.text`.
* I fixed the line-wrap on my commit message (and squashed my commits).

Please let me know if anything looks amiss!

timabbott commented:
OK, awesome, merged!

I'd definitely be happy to hear any suggestions for further things we can do to improve the experience and onboarding around notifications now that we've done this.

graue commented:
Maybe use heuristics to detect new users who aren't yet in the habit of checking Zulip, and show them (only) more notifications. You could also batch. Wait till a few messages come in, then be like: "See 5 unread messages in 'code > desktop notifications'". Once people are checking back often, stop sending them notifications of this type.
